File: mobile/test/p243_bos_durum_rehberi_test.dart
/// (P243 §6c) BOS DURUM MESAJLARI REHBER OLMALI — MOBIL.
///
/// =========================================================================
/// OLCULEN SEY
/// =========================================================================
/// Brief: "Boş liste ekranlarında 'Kayıt yok' yerine ne yapması
/// gerektiğini söyleyen mesaj olsun. TÜM boş ekranları tara ve düzelt."
///
/// Tarama iki ayri kusur buldu:
///   1. `BosDurum` kullanan 9 yerin 2'si `aciklama` VERMIYORDU,
///   2. 9 ekran `BosDurum`u HIC kullanmiyordu: ortasinda tek satir
///      `Center(child: Text(l10n.xxxYok))` vardi. Bu, bilesenin
///      var olmasina ragmen kuralin uygulanmadigi yerdi ve yalnizca
///      `BosDurum`u taramak onlari GORMEZDI.
///
/// Bu yuzden kilit IKI SEYI birden olcer: her `BosDurum` bir aciklama
/// tasir VE bos-liste dalinda ciplak `Center(child: Text(...))` kalmaz.
library;

import 'dart:io';

import 'package:flutter_test/flutter_test.dart';

/// `BosDurum(` cagrilarinin govdesini (parantez sayarak) ayikla.
List<String> _bosDurumGovdeleri(String kaynak) {
  final govdeler = <String>[];
  for (final e in RegExp(r'\bBosDurum\(').allMatches(kaynak)) {
    var i = e.end;
    var derinlik = 1;
    while (i < kaynak.length && derinlik > 0) {
      if (kaynak[i] == '(') derinlik++;
      if (kaynak[i] == ')') derinlik--;
      i++;
    }
    govdeler.add(kaynak.substring(e.end, i));
  }
  return govdeler;
}

Iterable<File> _dartDosyalari(Directory kok) =>
    kok.listSync(recursive: true).whereType<File>().where(
      (f) => f.path.endsWith('.dart'),
    );

void main() {
  final kok = Directory('lib/src');

  test('HER BosDurum bir aciklama tasir', () {
    final suclular = <String>[];
    for (final f in _dartDosyalari(kok)) {
      // Bilesenin KENDI tanimi haric: orada `aciklama` bir alandir.
      if (f.path.endsWith('core/ui/bos_durum.dart')) continue;
      for (final govde in _bosDurumGovdeleri(f.readAsStringSync())) {
        if (!govde.contains('aciklama:')) {
          suclular.add('${f.path}: ${govde.trim().split('\n').first}');
        }
      }
    }
    expect(suclular, isEmpty);
  });

  test('BOS LISTE dalinda CIPLAK Center(child: Text(...)) kalmadi', () {
    // Hata ve yukleme dallari disarida: orada tek satirlik metin dogru
    // cizimdir. Aranan sey "...Yok" / "...Bos" adli bir sozluk
    // anahtarinin ciplak bir `Text` icinde gecmesi.
    final desen = RegExp(
      r'Center\(\s*child:\s*Text\(\s*l10n\.\w*(Yok|Bos)\b',
      caseSensitive: true,
    );
    final suclular = <String>[];
    for (final f in _dartDosyalari(kok)) {
      final kaynak = f.readAsStringSync();
      for (final e in desen.allMatches(kaynak)) {
        suclular.add('${f.path}: ${e.group(0)}');
      }
    }
    expect(suclular, isEmpty);
  });

  test('TARAMA GERCEKTEN CALISIYOR (kilit kendini olcer)', () {
    // Kilit, kirilmadigi surece gecerli sayilmaz. Sahte kaynak uzerinde:
    // ic ice parantezli bir cagri govdesi dogru ayiklaniyor mu?
    final govdeler = _bosDurumGovdeleri('''
      BosDurum(ikon: Icons.a, baslik: l10n.b)
      BosDurum(ikon: Icons.c, baslik: f(g(1)), aciklama: l10n.d)
    ''');
    expect(govdeler, hasLength(2));
    expect(govdeler.where((g) => !g.contains('aciklama:')), hasLength(1));
  });
}

File: admin-web/tests/p243-bos-durum-rehberi.test.ts
// (P243 §6c) BOS DURUM MESAJLARI REHBER OLMALI.
//
// =========================================================================
// OLCULEN SEY
// =========================================================================
// Brief: "Boş liste ekranlarında 'Kayıt yok' yerine ne yapması
// gerektiğini söyleyen mesaj olsun. TÜM boş ekranları tara ve düzelt."
//
// Tarama yapildi: panelde 58 `BosDurum` kullanimi vardi ve 42'sinde
// YALNIZCA BASLIK vardi — yani ekran "Kayit yok" deyip susuyordu.
// Yeni bir tesiste bu ekranlarin cogu BOS ACILIR; kullanicinin ilk
// sorusu "bozuk mu, ben mi yapmadim" olur ve baslik bunu yanitlamaz.
//
// =========================================================================
// NEDEN METIN TARAMASI, DOM TESTI DEGIL
// =========================================================================
// 42 ekrani tek tek cizen 42 test yazmak ayni kurali 42 kez tekrar
// etmek olurdu ve YENI eklenen bir ekrani yine kacirirdi. Kural
// YAPISALDIR: "her `BosDurum` bir `aciklama` alir". Tarama, bugun
// yazilmamis bir ekrani da yarin yakalar.
import { readFileSync, readdirSync, statSync } from "node:fs";
import { join } from "node:path";

import { describe, expect, it } from "vitest";

const KOK = join(__dirname, "..");

function tsxDosyalari(dizin: string, biriktir: string[] = []): string[] {
  for (const ad of readdirSync(dizin)) {
    if (ad === "node_modules" || ad === ".next") continue;
    const yol = join(dizin, ad);
    if (statSync(yol).isDirectory()) tsxDosyalari(yol, biriktir);
    else if (ad.endsWith(".tsx")) biriktir.push(yol);
  }
  return biriktir;
}

/** `<BosDurum ...>` acilis etiketlerini (suslu parantezleri sayarak) ayikla. */
function bosDurumEtiketleri(kaynak: string): string[] {
  const etiketler: string[] = [];
  const desen = /<BosDurum\b/g;
  let eslesme: RegExpExecArray | null;
  while ((eslesme = desen.exec(kaynak)) !== null) {
    let i = eslesme.index;
    let derinlik = 0;
    while (i < kaynak.length) {
      const c = kaynak[i];
      if (c === "{") derinlik += 1;
      else if (c === "}") derinlik -= 1;
      else if (c === ">" && derinlik === 0) break;
      i += 1;
    }
    etiketler.push(kaynak.slice(eslesme.index, i + 1));
  }
  return etiketler;
}

describe("(P243 §6c) bos durum rehberi", () => {
  it("HER BosDurum bir `aciklama` tasir", () => {
    const suclular: string[] = [];
    for (const yol of [
      ...tsxDosyalari(join(KOK, "app")),
      ...tsxDosyalari(join(KOK, "components")),
    ]) {
      const kaynak = readFileSync(yol, "utf8");
      for (const etiket of bosDurumEtiketleri(kaynak)) {
        // Bilesenin KENDI tanimi haric (orada `aciklama` bir prop).
        if (yol.endsWith("durumlar.tsx")) continue;
        if (!/\baciklama=/.test(etiket)) {
          suclular.push(`${yol.slice(KOK.length + 1)} :: ${etiket.slice(0, 80)}`);
        }
      }
    }
    expect(suclular).toEqual([]);
  });

  it("TARAMA GERCEKTEN CALISIYOR (kilit kendini olcer)", () => {
    // Bir kilit, kirilmadigi surece gecerli sayilmaz. Burada sahte bir
    // kaynak uzerinde taranir: desen `aciklama`siz etiketi gorebiliyor
    // mu, ve suslu parantez icindeki `>` karakterini yutmuyor mu?
    const sahte = `
      <BosDurum baslik={t("a")} />
      <BosDurum baslik={n > 0 ? t("b") : t("c")} aciklama={t("d")} />
    `;
    const etiketler = bosDurumEtiketleri(sahte);
    expect(etiketler).toHaveLength(2);
    expect(etiketler.filter((e) => !/\baciklama=/.test(e))).toHaveLength(1);
  });
});
